plugin-server/src/worker/ingestion/event-pipeline/runner.ts

```diff
@@ -80,116 +50,164 @@ export class EventPipelineRunner {
-                } else {
-                    this.hub.statsd?.increment('kafka_queue.event_pipeline.step.last', {
-                        step: currentStepName,
-                        team_id: String(this.originalEvent?.team_id),
-                    })
-                    return {
-                        lastStep: currentStepName,
-                        args: currentArgs.map((arg: any) => this.serialize(arg)),
-                    }
-                }
-            } catch (error) {
-                await this.handleError(error, currentStepName, currentArgs)
-                return {
-                    lastStep: currentStepName,
-                    args: currentArgs.map((arg: any) => this.serialize(arg)),
-                    error: error.message,
-                }
+        if (processedEvent == null) {
+            return this.registerLastStep('pluginsProcessEventStep', event.team_id, [event])
+        }
+        const [normalizedEvent, newPersonContainer] = await this.runStep(processPersonsStep, [
+            this,
+            processedEvent,
+            personContainer,
+        ])
+
+        const preparedEvent = await this.runStep(prepareEventStep, [this, normalizedEvent])
+
+        await this.runStep(createEventStep, [this, preparedEvent, newPersonContainer])
+        return this.registerLastStep('createEventStep', event.team_id, [preparedEvent, newPersonContainer])
+    }
+
+    async runAsyncHandlersEventPipeline(event: PostIngestionEvent): Promise<EventPipelineResult> {
+        try {
+            this.hub.statsd?.increment('kafka_queue.event_pipeline.start', { pipeline: 'asyncHandlers' })
+            const personContainer = new LazyPersonContainer(event.teamId, event.distinctId, this.hub)
+            await this.runStep(runAsyncHandlersStep, [this, event, personContainer], false)
+            this.hub.statsd?.increment('kafka_queue.async_handlers.processed')
+            return this.registerLastStep('runAsyncHandlersStep', event.teamId, [event, personContainer])
+        } catch (error) {
+            if (error instanceof DependencyUnavailableError) {
+                // If this is an error with a dependency that we control, we want to
+                // ensure that the caller knows that the event was not processed,
+                // for a reason that we control and that is transient.
+                throw error
             }
+
+            return { lastStep: error.step, args: [], error: error.message }
         }
     }
 
-    protected runStep<Step extends StepType, ArgsType extends StepParameters<EventPipelineStepsType[Step]>>(
-        name: Step,
-        ...args: ArgsType
-    ): Promise<StepResult> {
+    registerLastStep(stepName: string, teamId: number | null, args: any[]): EventPipelineResult {
+        this.hub.statsd?.increment('kafka_queue.event_pipeline.step.last', {
+            step: stepName,
+            team_id: new String(teamId).toString(),
```

Comment:
Could this cause cardinality problems? Why do we have team_id in some places and not in others? If it's cool to have it, why not everywhere, if we shouldn't then I assume we shouldn't anywhere. Though I do see it was like this before, so feel free to ignore in this PR.

Reply:
I think I was just implementing what was already there but do say if not. I didn't want to change that functionality in this PR as it's already large.

Reply:
Have added a note that this might be high cardinality.